Fix connection aliasing.

markstory opened the pull request:
Revert 1162121f1e8b8d0f1df0fb65ddf8ace5af8693d0 Which caused non-default connections to be incorrectly aliased.

I've also updated the alias() doc blocks to make the parameter names a bit clearer and hopefully reduce confusion later. The original FriendsOfCake/Crud issue sounded like the reporter was missing a `test_alpha` connection.

Refs #8319

markstory committed:
Fix connection aliasing.

Revert 1162121f1e8b8d0f1df0fb65ddf8ace5af8693d0 Which caused non-default
connections to be incorrectly aliased.

I've also updated the alias() doc blocks to make the parameter names
a bit clearer and hopefully reduce confusion later. The original
FriendsOfCake/Crud issue sounded like the reporter was missing
a `test_alpha` connection.

Refs #8319

ADmad commented:
> The original FriendsOfCake/Crud issue sounded like the reporter was missing a test_alpha connection.

Does that mean one must always have a `test_*` connection for every non-default connection when trying to run tests?

ADmad commented:
Tests failing.

markstory commented:
> Does that mean one must always have a test_* connection for every non-default connection when trying to run tests, even if not actually using it in tests?

Yes. Right now the 'live' connection is used. I don't think we want that. I think for test suites to be representative models should use separate connections in test, when they use separate connections in the live mode. Another option would be to fallback to 'test' if no `test_stats` connection exists for `stats`.

ADmad commented:
Ensuring a "live" connection isn't used for tests does make sense. Guess we just need to iron out the implementation.

jadb commented:
Falling back would assume there are no conflicting table names, requiring another check or maybe the PDO exception raised could handle describing the issue. 

Another route would be to raise a missing connection exception instead and force uses to set a test connection (which is best practice I believe). Thoughts?

markstory committed:
Remove use of undefined connection name.

This custom connection was not defined and was causing missing
connection errors now that aliasing is correct. Instead of checking the
connection name this test can do its work by checking the object
instance type.

ADmad commented on tests/TestCase/ORM/Behavior/TranslateBehaviorTest.php:
```diff
@@ -92,8 +92,8 @@ public function testCustomTranslationTable()
         $this->assertEquals('\TestApp\Model\Table\I18nTable', $i18n->name());
+        $this->assertInstanceOf('TestApp\Model\Table\I18nTable', $i18n->target());
         $this->assertEquals('custom_i18n_table', $i18n->target()->table());
-        $this->assertEquals('test_custom_i18n_datasource', $i18n->target()->connection()->configName());
```
While checking connection name might not be necessary for this particular test, [test_custom_i18n_datasource](https://github.com/cakephp/cakephp/blob/master/tests/bootstrap.php#L100) is defined in tests/bootstrap.php

markstory replied:
I guess that could be deleted now too?

ADmad replied:
Guess so.

markstory commented:
> Another route would be to raise a missing connection exception instead and force uses to set a test connection (which is best practice I believe). Thoughts?

That is what happens now. If a `test_*` connection is missing an exception is raised when it is first used.

markstory commented:
@lorenzo I'm not sure this is entirely fixed. I'm going to dig a bit more. Something is not right as the generated connection map isn't always correct. @ADmad 's comment made me look at the existing code deeper and the default connection name in the I18nTable shouldn't have failed like it did.

ADmad commented:
> ..and the default connection name in the I18nTable shouldn't have failed like it did.

That's what I was thinking too. My understanding is that the aliasing done in FixtureManager is meant to alias `test_foo` as `foo` so that any app code which uses `foo` connection would actually use `test_foo` thereby ensuring "live" connection used. I believe the change I did correctly did that.

The issue reported in #8319 probably occurs because `stats` connection gets created first and hence aliasing `test_stats` as `stats` doesn't take effect.

markstory commented:
Aliases override concrete connections. The problematic commit made the `$map` created in FixtureManager to be inconsistent depending on whether the live or test connection was defined first in the datasources configuration.

ADmad commented:
> Aliases override concrete connections.

Okay, good to know, I wasn't sure of that.
